### PR TITLE
Allow Average Pooling modules to accept tensors with 0-dim batch sizes.

### PR DESCRIPTION
**Pull request by v0dro**
This issue fixes a part of https://github.com/pytorch/pytorch/issues/12013, which is summarized concretely in  https://github.com/pytorch/pytorch/issues/38115.

It introduces changes and tests for allowing the Average Pooling layers to accept tensors with 0 sized batch dimensions and return meaningful results.

### CODE REVIEW

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 1]
```diff
@@ -460,13 +461,16 @@ namespace {
...
         if (!output.is_contiguous(at::MemoryFormat::ChannelsLast)) {
           // TODO: modify this after resize_ added `memory_format` tag
           output.resize_({sizeB, sizeC, osizeH, osizeW}).as_strided_({sizeB, sizeC, osizeH, osizeW}, {sizeC*osizeH*osizeW, 1, osizeW*sizeC, sizeC});
         }
 
+        if (output.numel() == 0) {
+          return;
+        }
+
         const int max_threads = std::min<int>(
```
Comment (bjuncek) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:474]: *beginner question*
Why was this missing from the previous version? It seems like a fairly critical check no?

**Review by bjuncek — APPROVED** [review]
Looks good to me as most of the changes are simply a logical statements.

I've left two comments just because I haven't been writing tests in a while so I might be a bit rusty on the semantics

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu (hunk 1 above)
Comment (v0dro) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:474]: The previous version was designed to error out when the number of elements in the `output` are 0. Since this one is supposed to return a valid output, we use the if-check.

File: test/test_nn.py [hunk 2]
```diff
@@ -13258,6 +13258,25 @@ def test_MarginLoss_empty(self, device, dtype):
                 y = torch.ones(10, 0, device=device).type(torch.long)
                 mod(x, y)
 
+    @dtypes(torch.float, torch.double)
+    def test_adaptive_pooling_zero_batch(self, dtype):
+        # The device parameter does not accept device objects.
+        device_list = ['cpu']
+        if TEST_CUDA:
+            device_list.append('cuda')
```
Comment (krshrimali) [inline test/test_nn.py:13266]: Curious to know what's happening here, does:

```python
@onlyOnCPUAndCUDA
@dtypes(torch.float, torch.double)
def test_adaptive_pooling_zero_batch(self, dtype, device):
    # ...
```

not work? Apologies if this was answered by `The device parameter does not accept device objects` comment but I didn't completely understand this, hence the question.

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 3]
```diff
@@ -343,7 +343,8 @@ void adaptive_avg_pool3d_out_cuda_template(
 
   TORCH_CHECK(
       (input_.ndimension() == 4 || input_.ndimension() == 5),
-      "non-empty 4D or 5D (batch mode) tensor expected for input");
+      "Expected non-empty 4D or 5D vector or matrix with optional 0-dim batch size, but got",
```
Comment (krshrimali) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:346]: Do you want to add a space after `got` for better readability?

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 4]
```diff
@@ -589,7 +598,8 @@ namespace {
       case at::MemoryFormat::ChannelsLast: {
         // special case for tensor memory format in channels_last
         TORCH_CHECK(input.ndimension() == 4,
-          "non-empty 4D (batch mode) tensor expected for input with channels_last layout");
+                    "Expected non-empty vector or matrix with optional 0-dim batch size, but got",
```
Comment (krshrimali) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:601]: Same question, if you want to add a space after `got` here for readability reasons?

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 5]
```diff
@@ -527,7 +531,8 @@ namespace {
       }
       case at::MemoryFormat::Contiguous: {
         TORCH_CHECK((input.ndimension() == 3 || input.ndimension() == 4),
-          "non-empty 3D or 4D (batch mode) tensor expected for input");
+                    "Expected non-empty 3D or 4D vector or matrix with optional 0-dim batch size, but got",
```
Comment (krshrimali) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:534]: Same question as before on adding a space after `got`.

**Review by krshrimali — APPROVED** [review]
LGTM, thanks @v0dro! I had a small question about taking `device` as a parameter + a minor nit on adding space for the error messages.

File: test/test_nn.py (hunk 2 above)
Comment (v0dro) [reply test/test_nn.py:13266]: Yeah the module code seems to be outdated and only accepts devices in string form.

Comment (krshrimali) [reply test/test_nn.py:13266]: Interesting, thanks for clarifying @v0dro! :)

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu (hunk 1 above)
Comment (bjuncek) [reply aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:474]: ah, makes sense. Thanks

**Comment by v0dro** [comment]
@jbschlosser could you please review this PR?

**Comment by v0dro** [comment]
Ping @jbschlosser!

**Comment by rgommers** [comment]
> Ping @jbschlosser!

Hi @v0dro, please remember to actually request a review (via the _Reviewers_ field, not a comment). 10 days seems reasonable to send a gentle ping for another review (unless @jbschlosser indicates otherwise), but if you don't request the review it's not in his review queue to begin with.

**Comment by jbschlosser** [comment]
Hey @v0dro, sorry for the delay! As @rgommers mentioned, it's certainly helpful to add it to my review queue by requesting me as reviewer.

File: test/test_nn.py (hunk 2 above)
Comment (jbschlosser) [reply test/test_nn.py:13308]: I don't think this is true? Test works for me as written above by @krshrimali

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! :) Added some nitpicky comments below for consistency with some similar PRs

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 6]
```diff
@@ -446,7 +446,8 @@ namespace {
       case at::MemoryFormat::ChannelsLast: {
         // special case for tensor memory format in channels_last
         TORCH_CHECK(input.ndimension() == 4,
-          "non-empty 4D (batch mode) tensor expected for input with channels_last layout");
+                    "adaptive_avg_pool2d(): Expected 4D tensor, but got",
```
Comment (jbschlosser) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:449]: formatting nit: need a space after "got"

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 7]
```diff
@@ -527,7 +531,8 @@ namespace {
       }
       case at::MemoryFormat::Contiguous: {
         TORCH_CHECK((input.ndimension() == 3 || input.ndimension() == 4),
-          "non-empty 3D or 4D (batch mode) tensor expected for input");
+                    "adaptive_avg_pool2d(): Expected 3D or 4D tensor, but got",
```
Comment (jbschlosser) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:534]: formatting nit: need a space after "got"

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu [hunk 8]
```diff
@@ -589,7 +598,7 @@ namespace {
       case at::MemoryFormat::ChannelsLast: {
         // special case for tensor memory format in channels_last
         TORCH_CHECK(input.ndimension() == 4,
-          "non-empty 4D (batch mode) tensor expected for input with channels_last layout");
+                    "adaptive_avg_pool2d_backward_cuda(): Expected 4D tensor, but got", input.ndimension());
```
Comment (jbschlosser) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu:601]: formatting nit: need a space after "got"

File: aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu [hunk 9]
```diff
@@ -333,17 +333,17 @@ void adaptive_avg_pool3d_out_cuda_template(
...
-        "adaptive_avg_pool3d_cuda(): expected input to have non-empty spatial dimensions, "
+        "adaptive_avg_pool3d_cuda(): Expected input to have non-zero size for non-batch dimensions, "
         "but input has sizes ", input_.sizes(),
         " with dimension ", i, " being empty");
   }
 
   TORCH_CHECK(
       (input_.ndimension() == 4 || input_.ndimension() == 5),
-      "non-empty 4D or 5D (batch mode) tensor expected for input");
+      "adaptive_avg_pool3d_cuda(): Expected 4D or 5D tensor, but got", input_.sizes());
```
Comment (jbschlosser) [inline aten/src/ATen/native/cuda/AdaptiveAveragePooling3d.cu:346]: formatting nit: need a space after "got"

**Review by jbschlosser — COMMENTED** [review]
Looking great! Just a couple minor formatting tweaks needed for the error messages

**Comment by jbschlosser** [comment]
@v0dro FYI I'll be out next week- if the various 0 batch size PRs don't get finished up this week, I'll address them when I return 08/23

**Comment by v0dro** [comment]
@jbschlosser sorry for the delay. I have pushed the requested changes to all the PRs.  Hope we can merge all of them this week itself!

File: test/test_nn.py [hunk 10]
```diff
@@ -13324,6 +13324,27 @@ def test_MarginLoss_empty(self, device, dtype):
                 y = torch.ones(10, 0, device=device).type(torch.long)
                 mod(x, y)
 
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float, torch.double)
+    def test_adaptive_pooling_zero_batch(self, dtype):
```
Comment (jbschlosser) [inline test/test_nn.py:13329]: I think this should accept a `device` kwarg used below instead of creating & iterating over the `device_list`. This works locally for me, but let me know if you run into problems

Comment (v0dro) [reply test/test_nn.py:13329]: I used `@onlyOnCPUAndCUDA` and passed a device argument, but the code failed. You can see that other tests using adaptive pooling such as this one: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L5161 use the same approach.

File: test/test_nn.py (hunk 2 above)
Comment (v0dro) [reply test/test_nn.py]: Ah turns out it works on the latest master. I've changed it and pushed the code now.

**Review by jbschlosser — APPROVED** [review]
LGTM
